lib/smart_answer/state_resolver.rb

```diff
@@ -0,0 +1,85 @@
+module SmartAnswer
+  class StateResolver
+    def initialize(flow)
+      @flow = flow
+    end
+
+    def state_from_response_store(response_store, requested_node = nil)
+      forwarding_responses = @flow.response_store == :session ? {} : response_store.all
+      establish_response_store_state(start_state(forwarding_responses),
+                                     response_store,
+                                     requested_node)
+    end
+
+    def state_from_params(params)
+      responses = params[:responses].to_s.split("/")
+
+      state = responses.inject(start_state) do |current_state, response|
+        return current_state if current_state.error
+
+        transition_state_to_next_node(current_state, response)
+      end
+
+      if params[:next]
+        transition_state_to_next_node(state, params[:response])
+      elsif params[:previous_response]
+        apply_response_to_state(state, params[:previous_response])
+      else
+        state
+      end
+    end
+
+  private
+
+    def establish_response_store_state(state, response_store, requested_node)
+      node_name = state.current_node.to_s
+      no_response_for_question = !response_store.has?(node_name)
+      reached_an_outcome = @flow.node(node_name.to_sym).outcome?
+
+      return state if no_response_for_question || reached_an_outcome || state.error
+
+      response = response_store.get(node_name)
+      specifically_requested_node = requested_node == node_name
+
+      if specifically_requested_node
+        apply_response_to_state(state, response)
+      else
+        next_state = transition_state_to_next_node(state, response)
+        establish_response_store_state(next_state, response_store, requested_node)
+      end
```

Comment:
Nit: This is fab - I really like the recursive version of this. Maybe to highlight exactly what the base cases could write it like this with an explicit return symbol:

```suggestion
      return apply_response_to_state(state, response) if requested_node == node_name
      
      next_state = transition_state_to_next_node(state, response)
      establish_response_store_state(next_state, response_store, requested_node)
```

Sorry, I realise I previously asked for an inline comment to explain the this, however now that the code can be written more succinctly - there are fewer lines and variable to worry about and it's easier to understand the actual code.

Reply:
No worries, it's amended now.